src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/JoinVerticals.cs

```diff
@@ -0,0 +1,290 @@
+            {
+                await DownloadArtifactFiles(
+                    BuildId,
+                    $"{verticalName}{_artifactNameSuffix}",
+                    addedPackageIds,
+                    addedBlobIds);
+            }
+        }
+
+        // Create MergedManifest.xml
+        // taking the attributes from the main manifest
+        XElement mainManifestRoot = verticalManifests.First().Root
+            ?? throw new ArgumentException("The root element of the vertical manifest is null.");
+        mainManifestRoot.Attribute(_verticalNameAttribute)!.Remove();
+
+        string manifestOutputPath = Path.Combine(OutputFolder, "MergedManifest.xml");
+        XDocument mergedManifest = new(new XElement(
+            mainManifestRoot.Name,
+            mainManifestRoot.Attributes(),
+            packageElements.Values.OrderBy(elem => elem.Attribute(_idAttribute)?.Value),
+            blobElements.Values.OrderBy(elem => elem.Attribute(_idAttribute)?.Value)));
+
+        File.WriteAllText(manifestOutputPath, mergedManifest.ToString());
+    }
+
+    /// <summary>
+    /// Downloads specified packages and symbols from a specific build artifact and stores them in an output folder, 
+    /// either flat or with the same relative path as in the artifact.
+    /// </summary>
+    private async Task DownloadArtifactFiles(
+        string buildId,
+        string artifactName,
+        List<string> packageFileNames,
+        List<string> assetFileNames)
+    {
+        string packagesOutputPath = Path.Combine(OutputFolder, _packagesFolderName);
+        string assetsOutputPath = Path.Combine(OutputFolder, _assetsFolderName);
+
+        using AzureDevOpsClient azureDevOpsClient = new(AzureDevOpsToken, AzureDevOpsBaseUri, AzureDevOpsProject, Log);
+
+        ArtifactFiles filesInformation = await azureDevOpsClient.GetArtifactFilesInformation(buildId, artifactName, _retryCount);
+
+        foreach (var package in packageFileNames)
+        {
+            await DownloadFileFromArtifact(filesInformation, artifactName, azureDevOpsClient, buildId, package, packagesOutputPath);
+        }
+
+        foreach (var asset in assetFileNames)
+        {
+            await DownloadFileFromArtifact(filesInformation, artifactName, azureDevOpsClient, buildId, asset, assetsOutputPath);
+        }
+    }
+
+    private async Task DownloadFileFromArtifact(ArtifactFiles artifactFilesMetadata, string azureDevOpsArtifact, AzureDevOpsClient azureDevOpsClient, string buildId, string manifestFile, string destinationDirectory)
+    {
+        ArtifactItem fileItem;
+
+        var matchingFilePaths = artifactFilesMetadata.Items.Where(f => Path.GetFileName(f.Path) == Path.GetFileName(manifestFile));
+
+        if (!matchingFilePaths.Any())
+        {
+            throw new ArgumentException($"File {manifestFile} not found in source files.");
+        }
+
+        if (matchingFilePaths.Count() > 1)
+        {
+            if (manifestFile.Contains("productVersion.txt"))
```

Comment:
Did we get an issue opened for this? If not, can you do so (in source-build is fine). Either way, can you add a comment and link to it here?

Reply:
yep - https://github.com/dotnet/source-build/issues/4596 , will link it

Reply:
added